lib/rules/no-alert.js

```diff
@@ -8,44 +10,123 @@
+ */
+function findReference(scope, node) {
+    var references = scope.references.filter(function (reference) {
+        return reference.identifier.range[0] === node.range[0] &&
+            reference.identifier.range[1] === node.range[1];
+    });
+
+    if (references.length === 1) {
+        return references[0];
+    }
 }
 
-function report(context, node, result) {
-    context.report(node, "Unexpected {{name}}.", { name: result[1] });
+/**
+ * Checks if the given identifier name is shadowed in the given global scope.
+ * @param {Object} globalScope The global scope.
+ * @param {string} identifierName The identifier name to check
+ * @returns {boolean} Whether or not the name is shadowed globally.
+ */
+function isGloballyShadowed(globalScope, identifierName) {
+    return globalScope.variables.some(function (variable) {
+        return variable.name === identifierName && variable.defs.length > 0;
+    });
 }
 
+/**
+ * Checks if the given identifier node is shadowed in the given scope.
+ * @param {Object} scope The current scope.
+ * @param {Object} globalScope The global scope.
+ * @param {string} node The identifier node to check
+ * @returns {boolean} Whether or not the name is shadowed.
+ */
+function isShadowed(scope, globalScope, node) {
+    var reference = findReference(scope, node),
+        identifierName = node.name;
+
+    if (reference) {
+        if (reference.resolved || isGloballyShadowed(globalScope, identifierName)) {
+            return true;
+        }
+    }
+
+    return false;
+}
 
 //------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
 module.exports = function(context) {
+    var globalScope;
 
     return {
 
-        "CallExpression": function(node) {
+        "Program": function () {
+            globalScope = context.getScope();
+        },
 
-            var result;
+        "CallExpression": function(node) {
+            var callee = node.callee,
+                identifierName,
+                scope;
 
             // without window.
-            if (node.callee.type === "Identifier") {
+            if (callee.type === "Identifier") {
+                identifierName = callee.name;
+                scope = context.getScope();
 
-                result = matchProhibited(node.callee.name);
-
-                if (result) {
-                    report(context, node, result);
+                if (!isShadowed(scope, globalScope, callee) && isProhibitedIdentifier(callee.name)) {
+                    report(context, node, identifierName);
                 }
 
-            } else if (node.callee.type === "MemberExpression" && node.callee.property.type === "Identifier") {
-
-                result = matchProhibited(node.callee.property.name);
+            } else if (callee.type === "MemberExpression" && callee.object.name === "window") {
```

Comment:
You don't know that this is a reference to the global object, though. I could have a global called `window` in my node application, and it could be a GUI window object, not a reference to the global object.